### PR TITLE
Compose RunIf utilities

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

The Trainer's `_RunIf` calls Fabric's `_RunIf`
Pytest is kept to the tests directory exclusively.

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

**Review by Borda — CHANGES_REQUESTED** [review]

File: src/lightning/pytorch/utilities/testing/_runif.py [hunk 1]
```diff
@@ -11,186 +11,53 @@
...
 from lightning.pytorch.core.module import _ONNX_AVAILABLE
-from lightning.pytorch.strategies.deepspeed import _DEEPSPEED_AVAILABLE
-from lightning.pytorch.utilities.imports import _IPU_AVAILABLE, _OMEGACONF_AVAILABLE
+from lightning.pytorch.utilities.imports import _OMEGACONF_AVAILABLE
 
 _SKLEARN_AVAILABLE = RequirementCache("scikit-learn")
 
 
 def _RunIf(
-    min_cuda_gpus: int = 0,
```
[the remark this answers is not in the document]
Comment (carmocca) [reply src/lightning/pytorch/utilities/testing/_runif.py:37]: Sorry, what do you mean?

[the remark this answers is not in the document]
Comment (Borda) [reply src/lightning/pytorch/utilities/testing/_runif.py:37]: It is not clear to a user that he can use these arguments as they are given in other functions for IDE hints that are silent... in such cases can utilize both `RunIf`

```py
@BaseRunIf(min_torch='1.12', ...)  # <- Fabric version
@ExtendRunIf(psutil=True, ...)  # <- Pytorch version
def test_anytning_you_want(...)

```

File: src/lightning/fabric/utilities/testing/_runif.py [hunk 2]
```diff
@@ -41,19 +40,9 @@ def _RunIf(
     standalone: bool = False,
     deepspeed: bool = False,
     dynamo: bool = False,
-    **kwargs: Any,
```
Comment (Borda) [inline src/lightning/fabric/utilities/testing/_runif.py:44]: why no kwasrgs?

File: src/lightning/pytorch/utilities/testing/_runif.py [hunk 3]
```diff
@@ -11,186 +11,53 @@
...
-            # AssertionError: Torch not compiled with CUDA enabled
-            # RuntimeError: Found no NVIDIA driver on your system.
-            is_unrelated = "Found no NVIDIA driver" not in str(ex) or "Torch not compiled with CUDA" not in str(ex)
-            if is_unrelated:
-                raise ex
-            cond = True
 
-        conditions.append(cond)
-        reasons.append("CUDA device bf16")
+    reasons, kwargs = FabricRunIf(**kwargs)
```
Comment (Borda) [inline src/lightning/pytorch/utilities/testing/_runif.py:46]: here IDE hint wot help with suggesting the base args

File: src/lightning/pytorch/utilities/testing/_runif.py [hunk 4]
```diff
@@ -11,186 +11,53 @@
...
-            cond = sys.platform == "win32" or sys.version_info >= (3, 11)
-
-        # set use_base_version for nightly support
-        cond |= compare_version("torch", operator.lt, "2.0.0", use_base_version=True)
-        conditions.append(cond)
-        reasons.append("torch.dynamo")
-
-    reasons = [rs for cond, rs in zip(conditions, reasons) if cond]
-    kwargs.pop("condition", None)
-    kwargs.pop("reason", None)
```
Comment (Borda) [inline src/lightning/pytorch/utilities/testing/_runif.py:195]: this should stay

File: src/lightning/fabric/utilities/testing/_runif.py (hunk 2 above)
Comment (carmocca) [reply src/lightning/fabric/utilities/testing/_runif.py:44]: It's unused now

File: src/lightning/pytorch/utilities/testing/_runif.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply src/lightning/pytorch/utilities/testing/_runif.py:37]: Added back

File: src/lightning/pytorch/utilities/testing/_runif.py (hunk 4 above)
Comment (carmocca) [reply src/lightning/pytorch/utilities/testing/_runif.py:195]: There's no kwargs anymore

**Review by awaelchli — APPROVED** [review]
I'm suspicious of the fact that this runif gets refactored so often. Now there are four different RunIf. The previous version was definitely easy to understand and adapt for other projects.

**Comment by Borda** [comment]
> I'm suspicious of the fact that this runif gets refactored so often. Now there are four different RunIf. The previous version was definitely easy to understand and adapt for other projects.

I did not change much (I believe) on the internals, mostly moved it to the package to be re-used...

**Pull request merged**